### PR TITLE
Upload resume

### PR DESCRIPTION
**Pull request by SJLC**
### resume option for collections upload

The new --resume-media option will first check whether a file queued up
to be uploaded has previously been uploaded to Conservator in the 
destination folder (files of the same name but different folders are 
ignored).

* If there is an incomplete upload at the destination path,
  that invalid partial file will be deleted from Conservator
  before new upload attempt
* If there is a complete file at the destination path but it has 
  different md5sum than the local copy, that media file will be
  removed from the destination folder to avoid 2 files of same name
  in folder (but will not be altogether deleted from Conservator)
* If there is a complete file at the destination path that matches
  the md5sum of the local copy, upload of that file can be skipped

### CODE REVIEW

File: FLIR/conservator/cli/managers.py [hunk 1]
```diff
@@ -165,6 +165,11 @@ def download(
             is_flag=True,
             help="Include child directories recursively, creating collections as needed",
         )
+        @click.option(
+            "--resume-media",
+            is_flag=True,
+            help="Check if media files were previously uploaded, skip if so",
```
Comment (n-wach) [inline FLIR/conservator/cli/managers.py:171]: Do you think `--resume-media` should be the default behavior?

File: FLIR/conservator/conservator.py [hunk 2]
```diff
@@ -92,44 +87,6 @@ def default(save=True):
...
-        else:
-            collection_id = collection.id
-
-        upload_request = MediaUploadRequest(
-            file_path=file_path, collection_id=collection_id, remote_name=remote_name
-        )
-        result = MediaType.upload(self, upload_request)
-        return result.media_id
-
     def get_media_instance_from_id(self, media_id, fields=None):
```
Comment (n-wach) [inline FLIR/conservator/conservator.py:90]: Can this be moved to MediaManager too?

File: FLIR/conservator/conservator.py [hunk 3]
```diff
@@ -148,136 +105,6 @@ def get_media_instance_from_id(self, media_id, fields=None):
...
-class ProcessingTimeoutError(TimeoutError):
-    """
-    Raised when the amount of time spent waiting for media to process exceeds
-    the requested timeout.
-    """
-
-    pass
-
 
 class UnknownMediaIdException(Exception):
```
Comment (n-wach) [inline FLIR/conservator/conservator.py:109]: This too

File: FLIR/conservator/managers/__init__.py [hunk 4]
```diff
@@ -122,8 +121,9 @@ def upload(
 
         if media:
             logger.info("Uploading media to collection %s", collection.path)
-            self._conservator.upload_many_to_collection(
-                media_paths, collection, max_retries=max_retries
+            media_manager = MediaTypeManager(self._conservator)
```
Comment (n-wach) [inline FLIR/conservator/managers/__init__.py:124]: Can we add `MediaTypeManager` instance to `Conservator` in `__init__`?
For instance, `self.media = MediaTypeManager(self)`, so here you could do `media_manager = self._conservator.media`

File: FLIR/conservator/managers/media.py [hunk 5]
```diff
@@ -1,3 +1,257 @@
...
+        name = string.split("/")[-1]
+
+        parent = self._conservator.collections.from_remote_path(
+            path=parent_path, make_if_no_exist=False, fields="id"
+        )
+
+        # look inside parent for media with exact name match
+        fields = FieldsRequest.create(fields)
+        fields.include_field("name")
+        media = list(parent.get_media(fields=fields, search_text="name:" + name))
```
Comment (n-wach) [inline FLIR/conservator/managers/media.py:44]: PaginatedQuery can filter for exact name match

```suggestion
        media = list(parent.get_media(fields=fields, search_text="name:" + name).filtered_by(name=name))
```

File: FLIR/conservator/managers/searchable.py [hunk 6]
```diff
@@ -2,6 +2,11 @@
 from FLIR.conservator.paginated_query import PaginatedQuery
 
 
+class AmbiguousIdentifierException(Exception):
+    def __init__(self, identifier):
+        super().__init__(f"Multiple items found for '{identifier}', use ID")
```
Comment (n-wach) [inline FLIR/conservator/managers/searchable.py:7]: I think this should go alongside generic `TypeManager`

File: FLIR/conservator/util.py [hunk 7]
```diff
@@ -78,6 +78,16 @@ def download_file(path, name, url, silent=False, no_meter=False):
...
 
 
+def md5sum_file(path, block_size=1024 * 1024):
+    hasher = hashlib.md5()
+    with open(path, "rb") as fp:
+        block = fp.read(block_size)
+        while block:
+            hasher.update(block)
+            block = fp.read(block_size)
+    return hasher.hexdigest()
```
Comment (n-wach) [inline FLIR/conservator/util.py:88]: IIRC, `LocalDataset` also gets MD5 of local files when uploading staged images. Can you update that to use this utility?

File: FLIR/conservator/wrappers/collection.py [hunk 8]
```diff
@@ -160,56 +161,80 @@ def recursively_get_children(self, include_self=False, fields=None):
...
             self._conservator,
             Dataset,
             Query.datasets,
             fields=fields,
+            search_text=search_text,
             collection_id=self.id,
         )
         return datasets
 
+    def remove_media(self, media_id):
```
Comment (n-wach) [inline FLIR/conservator/wrappers/collection.py:225]: Might as well change this to `remove_medias` that takes a list of ID to remove; and make `remove_media` call with single ID

File: FLIR/conservator/wrappers/image.py [hunk 9]
```diff
@@ -1,8 +1,25 @@
...
+
+        Returns result as MediaCompare object
+
+        :param local_path: Path to local copy of file for comparison with remote.
+        """
+        result = MediaCompare.MISMATCH
+        local_md5 = md5sum_file(local_path)
+        if local_md5 == self.image_md5:
+            result = MediaCompare.MATCH
+        return result
```
Comment (n-wach) [inline FLIR/conservator/wrappers/image.py:25]: Very similar code in `Video`--can we extract to helper such as `MediaCompare.with_local_file(local_path, remote_md5)`?

```suggestion
        local_md5 = md5sum_file(local_path)
        if local_md5 == self.image_md5:
            return MediaCompare.MATCH
        return MediaCompare.MISMATCH
```

File: FLIR/conservator/managers/__init__.py (hunk 4 above)
Comment (n-wach) [reply FLIR/conservator/managers/__init__.py:124]: Hmm, I suppose `conservator.videos` and `images` are already `MediaTypeManager`, so maybe reuse one of those for now

File: FLIR/conservator/conservator.py (hunk 2 above)
Comment (n-wach) [reply FLIR/conservator/conservator.py:90]: On second thought it can stay until we have a generic `Conservator.media`

**Review by n-wach — CHANGES_REQUESTED** [review]

File: FLIR/conservator/util.py (hunk 7 above)
Comment (n-wach) [reply FLIR/conservator/util.py:88]: line 300

File: FLIR/conservator/cli/managers.py (hunk 1 above)
Comment (SJLC) [reply FLIR/conservator/cli/managers.py:171]: The overhead is an extra phase at the beginning of each folder querying for all the media in it. If you've got 1200 videos like in some projects that could be substantial. I figured somebody should know if they tried to upload a directory before or not, so they would know when to pass in resume.

That said, I wouldn't be averse to flipping the default. Thoughts?

File: FLIR/conservator/managers/__init__.py (hunk 4 above)
Comment (SJLC) [reply FLIR/conservator/managers/__init__.py:124]: I thought about having a default instance for media manager, but media is not a full-featured searchable graphql-query-backed structure like collection, dataset, video, or image, just a place to put common code supporting video and image. So I think API users would be disappointed to see this media instance and find that it is really not like the others, kind of ugly API design to imply it is similar. For that reason I'd rather use `videos` instance for upload.

File: FLIR/conservator/managers/searchable.py (hunk 6 above)
Comment (SJLC) [reply FLIR/conservator/managers/searchable.py:7]: ok

File: FLIR/conservator/util.py (hunk 7 above)
Comment (SJLC) [reply FLIR/conservator/util.py:88]: Good idea. Folder download is the other place I'm planning to use it.

File: FLIR/conservator/wrappers/image.py (hunk 9 above)
Comment (SJLC) [reply FLIR/conservator/wrappers/image.py:25]: Comparing a couple checksums isn't inherently a media-specific task; we might want to run validation on file-locker or metadata files later on. How about moving more into util, including a generically named enum "FileCompare".

File: FLIR/conservator/managers/media.py (hunk 5 above)
Comment (SJLC) [reply FLIR/conservator/managers/media.py:44]: This is an example of media not being so parallel to the other classes. The return of get_media() isn't actually a `PaginatedQuery` but rather a `generator` that will first iterate through results of the video query and then through results of the image query, so it's too late to add a filtered_by() at that point.

File: FLIR/conservator/wrappers/image.py (hunk 9 above)
Comment (SJLC) [reply FLIR/conservator/wrappers/image.py:25]: On the other hand Conservator isn't providing md5sum for file-locker or metadata files yet, so maybe it is media-specific

File: FLIR/conservator/managers/__init__.py (hunk 4 above)
Comment (n-wach) [reply FLIR/conservator/managers/__init__.py:122]: Makes sense. I think eventually we should create a "fully-featured searchable" manager that combines Video and Image paginated queries, and move a lot of this over there. 

We can also discuss with conservator folks if it's worth making new GraphQL Media queries that combine Videos and Images (shouldn't be hard. internally they're stored in the same db and filtered by `frameCount`). The conservator team has said they're willing to expand/create GraphQL queries for use cases outside of frontend UI

File: FLIR/conservator/managers/media.py (hunk 5 above)
Comment (n-wach) [reply FLIR/conservator/managers/media.py:44]: Ah yes you're right. Hopefully we can come up with a more elegant solution to the Media queries

File: FLIR/conservator/wrappers/image.py (hunk 9 above)
Comment (n-wach) [reply FLIR/conservator/wrappers/image.py]: Either way is fine with me

File: FLIR/conservator/cli/managers.py (hunk 1 above)
Comment (n-wach) [reply FLIR/conservator/cli/managers.py:171]: That's a good point, and I guess there are tradeoffs both ways. 

The current implementation is on a per-local-file basis, checking if collection has a match of the single name. We *could* fetch all media at once on a per-collection basis and compare names ourselves. Would likely be a significant speedup, would have minimal slowdown when uploading to an empty collection. 

I guess I'd slightly prefer switching to resume by default and optimizing, but am fine if you want to keep it as-is.

File: FLIR/conservator/wrappers/collection.py [hunk 10]
```diff
@@ -160,56 +161,80 @@ def recursively_get_children(self, include_self=False, fields=None):
...
 
-    def recursively_get_images(self, fields=None):
+    def recursively_get_images(self, fields=None, search_text=""):
         """Yields all images in this and child collections recursively"""
         for collection in self.recursively_get_children(include_self=True, fields="id"):
-            yield from collection.get_images(fields)
+            yield from collection.get_images(fields, search_text)
 
-    def get_videos(self, fields=None):
+    def get_videos(self, fields=None, search_text=""):
```
Comment (rlc2) [inline FLIR/conservator/wrappers/collection.py:181]: Nice feature for get_videos() and get_images() to have a search_text parameter :+1:

File: FLIR/conservator/wrappers/image.py [hunk 11]
```diff
@@ -1,8 +1,20 @@
 from FLIR.conservator.generated import schema
-from FLIR.conservator.wrappers.media import MediaType
+from FLIR.conservator.wrappers.media import MediaType, MediaCompare
```
Comment (rlc2) [inline FLIR/conservator/wrappers/image.py:2]: I don't actually see MediaCompare in the rest of the diff for this file?

File: FLIR/conservator/wrappers/video.py [hunk 12]
```diff
@@ -3,14 +3,26 @@
 from FLIR.conservator.generated.schema import Query, Mutation
 from FLIR.conservator.paginated_query import PaginatedQuery
 from FLIR.conservator.wrappers.frame import Frame
-from FLIR.conservator.wrappers.media import MediaType
+from FLIR.conservator.wrappers.media import MediaType, MediaCompare
```
Comment (rlc2) [inline FLIR/conservator/wrappers/video.py:6]: Same comment as in image.py; the verify_md5() method is what returns a MediaCompare object, but the source for this module doesn't instantiate the class itself.

**Review by rlc2 — APPROVED** [review]

File: FLIR/conservator/local_dataset.py [hunk 13]
```diff
@@ -292,7 +292,7 @@ def get_image_info(path):
             "width": image.width,
             "height": image.height,
             "fileSize": os.path.getsize(path),
-            "md5": hashlib.md5(open(path, "rb").read()).hexdigest(),
+            "md5": md5sum_file(path, "rb"),
```
Comment (n-wach) [inline FLIR/conservator/local_dataset.py:295]: Is the "rb" a typo?

File: FLIR/conservator/managers/media.py [hunk 14]
```diff
@@ -1,3 +1,257 @@
...
+        """
+        redo_upload_paths = file_paths
+        collection.populate("path")
+        # iterate copy of list since items might be removed
+        for path in list(file_paths):
+            (local_path, remote_name) = path
+            if remote_name is None:
+                remote_name = os.path.split(local_path)[-1]
+            remote_path = "/".join((collection.path, remote_name))
+            media = self.from_path(remote_path, fields="state")
```
Comment (n-wach) [inline FLIR/conservator/managers/media.py:207]: Potential for `AmbiguousIdException` here

File: FLIR/conservator/wrappers/collection.py (hunk 10 above)
Comment (n-wach) [reply FLIR/conservator/wrappers/collection.py:181]: definitely

File: FLIR/conservator/wrappers/media.py [hunk 15]
```diff
@@ -58,6 +68,16 @@ class MediaType(QueryableType, FileLockerType, MetadataType):
     metadata_gen_url = Mutation.generate_signed_metadata_upload_url
     metadata_confirm_url = Mutation.mark_annotation_as_uploaded
 
+    def verify_md5(self, local_path, expected_md5):
```
Comment (n-wach) [inline FLIR/conservator/wrappers/media.py:71]: `@staticmethod`

**Review by n-wach — COMMENTED** [review]
Looks good, just a few minor things

File: FLIR/conservator/managers/__init__.py (hunk 4 above)
Comment (SJLC) [reply FLIR/conservator/managers/__init__.py:122]: Well that would be nice for API users to have built-in support for media in Conservator! Yes let's mention it to them.

File: FLIR/conservator/wrappers/image.py (hunk 9 above)
Comment (SJLC) [reply FLIR/conservator/wrappers/image.py]: I went with your original suggestion of media for above reason

**Pull request merged**